Content.Shared/Attachments/Components/AttachmentHolderComponent.cs

```diff
@@ -0,0 +1,38 @@
+﻿using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Attachments.Components;
+
+[RegisterComponent]
+public sealed partial class AttachmentHolderComponent : Component
+{
+    /// <summary>
+    ///     Components this object will be granted.
+    /// </summary>
+    [ViewVariables]
+    [DataField]
+    public Dictionary<string, ComponentRegistry>? Components;
+
+    /// <summary>
+    ///     Prototype which we will pull components from if <see cref="Components"/>> is null.>.
+    /// </summary>
+    [ViewVariables]
+    [DataField]
+    public Dictionary<string, EntProtoId>? Prototypes;
+
+    /// <summary>
+    ///     Fields from components that will be inherited from the inserted item.
+    /// </summary>
+    [ViewVariables]
+    [DataField]
+    public Dictionary<string, List<string>>? Fields;
+
+    /// <summary>
+    ///     If a component name exists in this list, every field will be copied from the reference.
+    /// </summary>
+    [ViewVariables]
+    [DataField]
+    public List<string>? CopyAllFields;
+
+    [ViewVariables(VVAccess.ReadOnly)]
+    public List<(EntityUid, Type)> AddedComps = new ();
+}
```

Comment:
This is trying to be a godcomp way, way too much, at best it should be a componentregistry and leveraging the engine copycomps / having a method to restore them if desired.